File: java/src/main/java/org/twyszomirski/aoc/solutions/Day_2.java
package org.twyszomirski.aoc.solutions;

import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.function.Function;
import java.util.stream.Collectors;

import static java.io.IO.println;

public class Day_2 {

    public static void solve() throws IOException {
        var lines = Files.readAllLines(Paths.get("src/main/resources/input_day_2.txt"));
        println("======== Day 2 ===========");
        part_1(lines);
        part_2(lines);
    }

    public static void part_1(List<String> lines) {
        var count = lines.stream().map(it -> Arrays.stream(it.split(" ")).map(Long::valueOf).toList())
                .filter(Day_2::isGoodReport).count();

        println(count);
    }

    private static boolean isGoodReport(List<Long> report) {
        var direction = report.get(0) - report.get(1) > 0 ? 1 : -1;

        for (int i = 0; i < report.size() - 1; i++) {
            var distance = Math.abs(report.get(i) - report.get(i + 1));
            if (distance < 1 || distance > 3) {
                return false;
            }
            var currentDir = report.get(i) - report.get(i + 1) > 0 ? 1 : -1;
            if (direction != currentDir) {
                return false;
            }
        }
        return true;
    }

    public static void part_2(List<String> lines) {
        var count = lines.stream().map(it -> Arrays.stream(it.split(" ")).map(Long::valueOf).toList())
                .filter(Day_2::isGoodReportWithReplacement).count();

        println(count);
    }

    private static boolean isGoodReportWithReplacement(List<Long> report) {
        if (isGoodReport(report)) {
            return true;
        }

        for (int i = 1; i < report.size(); i++) {
            var newList = new ArrayList<>(report);
            newList.remove(i);
            if (isGoodReport(newList)) {
                return true;
            }
        }
        return false;
    }
}

File: kotlin/src/main/kotlin/org/twyszomirski/aoc/solutions/Day_2.kt
package org.twyszomirski.aoc.solutions

import java.io.File
import kotlin.math.abs
import kotlin.math.sign


class Day_2 {

    fun solve(){
        val lines = File("src/main/resources/input_day_2.txt").readLines()
        println("======== Day 2 ===========")
        part1(lines)
        part2(lines)
    }

    fun part1(input: List<String>) {
        val count = input.map { it.split(" ").map { it.toLong() }
        }.count { isGoodReport(it) }

        println(count)
    }

    fun part2(input: List<String>) {
        val count = input.map { it.split(" ").map { it.toLong() }
        }.count { isGoodReportWithRemovalOption(it) }

        println(count)
    }

    fun isGoodReportWithRemovalOption(report: List<Long>): Boolean {
        if(isGoodReport(report)) return true

        (0 .. report.size).forEach {
            val isGood = isGoodReport(report.filterIndexed { idx, _ -> idx != it })
            if(isGood) return true
        }
        return false
    }

    fun isGoodReport(report: List<Long>): Boolean {
        val direction = (report[0] - report[1]).sign
        (0..<report.size - 1).forEach {
            val distance = report[it] - report[it + 1]
            val distanceAbs = abs(distance)
            if (distanceAbs < 1 || distanceAbs > 3) {
                return false
            }
            if (direction != distance.sign) {
                return false
            }
        }
        return true
    }
}